install-validator/internal/kubernetes/client.go

```diff
@@ -0,0 +1,113 @@
+package kubernetes
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	v1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	ext "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/yaml"
+)
+
+type ClientInt interface {
+	GetCrd(name string) *v1.CustomResourceDefinition
+	ApplyCrd(crd v1.CustomResourceDefinition) error
+	ListCrds() error
+	ListResources(crd v1.CustomResourceDefinition) ([]interface{}, error)
+}
+
+type Client struct {
+	clientset *ext.Clientset
+	config    *rest.Config
+	crds      []v1.CustomResourceDefinition
+}
+
+func NewClient() (Client, error) {
+	c := Client{}
+	err := c.setConfig()
+	if err != nil {
+		return c, err
+	}
+	err = c.setClientSet()
+	if err != nil {
+		return c, err
+	}
+	return c, nil
+}
+
+func (c *Client) setClientSet() error {
+	clientset, err := ext.NewForConfig(c.config)
+
+	if err != nil {
+		return err
+	}
+	c.clientset = clientset
+
+	return nil
+}
+
+func (c *Client) setConfig() error {
+	config, err := rest.InClusterConfig()
+	if err == nil {
+		c.config = config
+		return nil
+	}
+
+	home, exists := os.LookupEnv("HOME")
+	if !exists {
+		home = "/root"
+	}
+	configPath := filepath.Join(home, ".kube", "config")
+	config, err = clientcmd.BuildConfigFromFlags("", configPath)
+	if err != nil {
+		return err
+	}
+	c.config = config
+	return nil
+}
+
+func (c *Client) GetCrd(name string) *v1.CustomResourceDefinition {
+	for _, s := range c.crds {
+		if s.Name == name {
+			return &s
+		}
+	}
+	return nil
+}
+
+func (c *Client) ApplyCrd(crd v1.CustomResourceDefinition) error {
+	oldCrd := c.GetCrd(crd.Name)
+	var err error
+	if oldCrd == nil {
+		_, err = c.clientset.ApiextensionsV1().CustomResourceDefinitions().Create(context.TODO(), &crd, metav1.CreateOptions{})
+	} else {
+		crd.ObjectMeta.ResourceVersion = oldCrd.ObjectMeta.ResourceVersion
+		_, err = c.clientset.ApiextensionsV1().CustomResourceDefinitions().Update(context.TODO(), &crd, metav1.UpdateOptions{})
+	}
+	return err
+}
+
+func (c *Client) ListCrds() error {
+	l, err := c.clientset.ApiextensionsV1().CustomResourceDefinitions().List(context.TODO(), metav1.ListOptions{})
+	c.crds = l.Items
+	return err
+}
+
+func (c *Client) ListResources(crd v1.CustomResourceDefinition) ([]interface{}, error) {
+	data, err := c.clientset.ApiextensionsV1beta1().RESTClient().Get().RequestURI(createURI(crd)).DoRaw(context.TODO())
```

Comment:
Is this the simplest API to list all CR's for a CRD type ?

cc @PavitraS @kacpersaw

Reply:
For typed client - yes.
Alternate method is to use dynamic client with GVR to get the same output. But in this case it's redundant.

Reply:
We looked at it with @kacpersaw - the only thing we came up with is to use dynamic client, which seems doesn't seem to simplify the code